File: python/rapier-testbed/rapier_testbed/examples3/voxels3.py
"""Port of examples3d/voxels3.rs.

A voxelized wavy floor (built with ``Collider.voxels_from_points``) with a
grid of dynamic primitives — balls, cuboids, cylinders, cones, capsules —
raining onto it. Each filled voxel renders as its own cube.

The Rust example also supports interactive voxel editing with the mouse
(add/remove voxels under the cursor); that relies on testbed ray-picking
hooks the Python testbed doesn't expose, so it's omitted here. Counts are
scaled down from the Rust demo to keep the Python viewer responsive.
"""
from __future__ import annotations

import math

import numpy as np

import rapier3d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Voxels"

# Grid resolution of the floor (Rust uses 200; scaled down for the viewer).
_N = 64
_VOXEL_SIZE_Y = 1.0
# How many voxels thick the floor is. The Rust demo samples only the surface;
# at 200×200 the per-cell slope is gentle enough that the single-voxel shell is
# watertight. At our lower resolution the surface steps by up to ~2 voxels
# between neighbors, which would leave gaps a falling object slips through, so
# we fill a few voxels deep to keep the floor solid.
_THICKNESS = 3


def _floor_samples():
    """Point cloud sampling a wavy surface, with raised walls at the edges."""
    pts = []
    n = _N
    for i in range(n):
        for j in range(n):
            y = (
                np.clip(math.sin(i / n * 10.0), -0.8, 0.8)
                * np.clip(math.cos(j / n * 10.0), -0.8, 0.8)
                * 16.0
            )
            # Fill downward from the surface so the floor is a solid slab.
            for d in range(_THICKNESS):
                pts.append((float(i), (y - d) * _VOXEL_SIZE_Y, float(j)))
            # Walls along the border so edge objects don't fall into the void.
            if i == 0 or i == n - 1 or j == 0 or j == n - 1:
                for k in range(1, 5):
                    pts.append((float(i), (y + k) * _VOXEL_SIZE_Y, float(j)))
    return np.asarray(pts, dtype=np.float32)


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    # Voxelized wavy floor (one static voxel-grid collider).
    samples = _floor_samples()
    voxel_size = (1.0, _VOXEL_SIZE_Y, 1.0)
    colliders.insert(rp.Collider.voxels_from_points(voxel_size, samples))

    # Floor bounds, to scatter the falling objects over the surface.
    mins = samples.min(axis=0)
    maxs = samples.max(axis=0)
    extents = (maxs - mins) * 0.75
    margin = (maxs - mins - extents) / 2.0

    # Dynamic primitives raining down (kind cycles per layer, like "Mixed").
    nik = 12
    nlayers = 5
    ball_radius = 0.5
    for i in range(nik):
        for j in range(nlayers):
            for k in range(nik):
                x = mins[0] + margin[0] + i * extents[0] / nik
                y = maxs[1] + j * 2.0
                z = mins[2] + margin[2] + k * extents[2] / nik
                body = rp.RigidBody.dynamic().translation(
                    (float(x), float(y), float(z))
                )
                h = bodies.insert(body)
                kind = j % 5
                if kind == 0:
                    co = rp.Collider.ball(ball_radius)
                elif kind == 1:
                    co = rp.Collider.cuboid(ball_radius, ball_radius, ball_radius)
                elif kind == 2:
                    co = rp.Collider.cylinder(ball_radius, ball_radius)
                elif kind == 3:
                    co = rp.Collider.cone(ball_radius, ball_radius)
                else:
                    co = rp.Collider.capsule_y(ball_radius, ball_radius)
                colliders.insert_with_parent(co, h, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    # Look at the middle of the floor from a corner.
    center = (maxs + mins) * 0.5
    testbed.look_at(
        (float(maxs[0]) + 30.0, float(maxs[1]) + 40.0, float(maxs[2]) + 30.0),
        (float(center[0]), float(center[1]), float(center[2])),
    )


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")

File: bindings/bevy_rapier/bevy_rapier3d/examples/voxels3.rs
use bevy::prelude::*;
use bevy_rapier3d::prelude::*;
use rapier3d::math::Pose;

fn main() {
    App::new()
        .insert_resource(ClearColor(Color::srgb(
            0xF9 as f32 / 255.0,
            0xF9 as f32 / 255.0,
            0xFF as f32 / 255.0,
        )))
        .add_plugins((
            DefaultPlugins,
            RapierPhysicsPlugin::<NoUserData>::default(),
            RapierDebugRenderPlugin::default(),
        ))
        .add_systems(Startup, (setup_graphics, setup_physics))
        .run();
}

pub fn setup_graphics(mut commands: Commands) {
    commands.spawn((
        Camera3d::default(),
        Transform::from_xyz(-100.0, 100.0, -100.0).looking_at(Vec3::new(0.0, 0.0, 0.0), Vec3::Y),
    ));
}

pub fn setup_physics(mut commands: Commands) {
    /*
     * Create a voxelized wavy floor.
     */
    let voxel_size = Vec3::new(1f32, 1f32, 1f32);
    let mut samples = vec![];
    let n = 200;
    for i in 0..n {
        for j in 0..n {
            let y = (i as f32 / n as f32 * 10.0).sin().clamp(-0.8, 0.8)
                * (j as f32 / n as f32 * 10.0).cos().clamp(-0.8, 0.8)
                * 16.0;

            samples.push(Vec3::new(
                i as f32 * voxel_size.x,
                y * voxel_size.y,
                j as f32 * voxel_size.z,
            ));

            if i == 0 || i == n - 1 || j == 0 || j == n - 1 {
                // Create walls so the object at the edge don’t fall into the infinite void.
                for k in 0..4 {
                    samples.push(Vec3::new(
                        i as f32 * voxel_size.x,
                        (y + k as f32) * voxel_size.y,
                        j as f32 * voxel_size.z,
                    ));
                }
            }
        }
    }
    let collider = Collider::voxels_from_points(voxel_size, &samples);
    let ground_position = Vec3::new(voxel_size.x / 2f32, 0.0, voxel_size.z / 2f32);
    let floor_aabb = collider.raw.compute_aabb(&Pose::IDENTITY);
    commands.spawn((Transform::from_translation(ground_position), collider));

    /*
     * Create dynamic objects to fall on voxels.
     */
    let extents = floor_aabb.extents() * 0.75;
    let margin = (floor_aabb.extents() - extents) / 2.0;
    let nik = 30;
    for i in 0..nik {
        for j in 0..5 {
            for k in 0..nik {
                let falling_objects = (j + i + k) % 3;

                let ball_radius = 0.5;
                let co = match falling_objects {
                    0 => Collider::ball(ball_radius),
                    1 => Collider::cuboid(ball_radius, ball_radius, ball_radius),
                    2 => Collider::capsule_y(ball_radius, ball_radius),
                    _ => unreachable!(),
                };
                commands.spawn((
                    Transform::from_xyz(
                        floor_aabb.mins.x + margin.x + i as f32 * extents.x / nik as f32
                            - ground_position.x,
                        floor_aabb.maxs.y + j as f32 * 2.0 - ground_position.y,
                        floor_aabb.mins.z + margin.z + k as f32 * extents.z / nik as f32
                            - -ground_position.z,
                    ),
                    RigidBody::Dynamic,
                    co,
                ));
            }
        }
    }
}
